src/charm.py

```diff
@@ -1006,9 +1091,140 @@ def restart_backup_service(self) -> None:
+
+            self.secrets[scope][Config.Secrets.SECRET_LABEL] = secret
+            self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL] = {key: value}
+            logging.debug(f"Secret {scope}:{key} published (as first). ID: {secret.id}")
+            peer_data.update({Config.Secrets.SECRET_INTERNAL_LABEL: secret.id})
+
+        return self.secrets[scope][Config.Secrets.SECRET_LABEL].id
+
+    def _update_juju_secrets_cache(self, scope: Scopes) -> None:
+        """Helper function to get Juju secret."""
+        peer_data = self._peer_data(scope)
+
+        if not peer_data.get(Config.Secrets.SECRET_INTERNAL_LABEL):
+            return
+
+        if Config.Secrets.SECRET_CACHE_LABEL not in self.secrets[scope]:
+            try:
+                # NOTE: Secret contents are not yet available!
+                secret = self.model.get_secret(id=peer_data[Config.Secrets.SECRET_INTERNAL_LABEL])
+            except SecretNotFoundError as e:
+                logging.debug(
+                    f"No secret found for ID {peer_data[Config.Secrets.SECRET_INTERNAL_LABEL]}, {e}"
+                )
+                return
+
+            logging.debug(f"Secret {peer_data[Config.Secrets.SECRET_INTERNAL_LABEL]} downloaded")
+
+            # We keep the secret object around -- needed when applying modifications
+            self.secrets[scope][Config.Secrets.SECRET_LABEL] = secret
+
+            # We retrieve and cache actual secret data for the lifetime of the event scope
+            self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL] = secret.get_content()
+
+    def _get_juju_secrets_cache(self, scope: Scopes):
+        return self.secrets[scope].get(Config.Secrets.SECRET_CACHE_LABEL)
+
+    def _juju_secret_get(self, scope: Scopes, key: str) -> Optional[str]:
+        """Helper function to get Juju secret."""
+        if not key:
+            return
+
+        self._update_juju_secrets_cache(scope)
+        secret_cache = self._get_juju_secrets_cache(scope)
+        if secret_cache:
+            secret_data = secret_cache.get(key)
+            if secret_data and secret_data != Config.Secrets.SECRET_DELETED_LABEL:
+                logging.debug(f"Getting secret {scope}:{key}")
+                return secret_data
+        logging.debug(f"No value found for secret {scope}:{key}")
+
+    def _juju_secret_remove(self, scope: Scopes, key: str) -> None:
+        """Remove a Juju 3.x secret."""
+        self._update_juju_secrets_cache(scope)
+
+        secret = self.secrets[scope].get(Config.Secrets.SECRET_LABEL)
+        if not secret:
+            logging.error(f"Secret {scope}:{key} wasn't deleted: no secrets are available")
+            return
+
+        secret_cache = self.secrets[scope].get(Config.Secrets.SECRET_CACHE_LABEL)
+        if not secret_cache or key not in secret_cache:
+            logging.error(f"No secret {scope}:{key}")
+            return
+
+        secret_cache[key] = Config.Secrets.SECRET_DELETED_LABEL
```

Comment:
I thought we agreed that deletion labels aren't needed anymore...

Do we keep them still?

Reply:
I mean I would really like to believe that https://bugs.launchpad.net/juju/+bug/2028094 is already in `stable`...

Reply:
For now I prefer to keep the code for both charms similar and do the required updates later in a separate PR